### PR TITLE
traceql: fix signed-int parsing to correctly handle math.MinInt64 (issue #4623)

### PR DESCRIPTION
**Pull request by Syedowais312**
# 🔧 Fix: Proper Handling of Signed Integers in TraceQL Lexer  
Closes: #4623

## Summary

This PR fixes incorrect parsing of signed integers in the TraceQL lexer.

Previously, expressions such as:

    { -9223372036854775808 }
    { +10 }
    { -5s }

were incorrectly parsed or caused unexpected failures due to improper unary sign handling and integer overflow behavior.

### ✔ 1. Correct signed integer parsing
- Supports unary `+` and `-` in valid contexts.
- Handles the `MinInt64` edge case (-9223372036854775808) safely.
- Validates numeric ranges to prevent silent overflows.

### ✔ 2. Correct signed duration parsing
Signed durations are now parsed consistently:

    -5s
    +10ms

using the existing `tryScanDuration` logic.

### ✔ 3. Fixes unary parsing context rules
Unary signs are now only allowed:
- at expression start,
- after `{`,
- after comparison operators,
- after commas,
- after unary `!`.

(Prevents incorrect parsing inside function argument lists.)

## Tests

Added and updated tests in `parse_test.go`:
- `{ -9223372036854775808 }`
- `{ +9223372036854775807 }`
- `{ -5s }`
- `{ +10ms }`

All related TraceQL tests pass.

## Why This Fix?

The previous logic failed in edge cases and mixed incorrectly with duration parsing.  
This fix is:
- simple,
- safe,
- correct,
- aligned with TraceQL grammar expectations.

## Result

TraceQL signed integer parsing is now robust, predictable, and fully tested.

## 🙏 Notes

If additional signed numeric types (float, duration, hex, etc.) need first-class support, I’d be happy to extend this work.

### CODE REVIEW

**Comment by Syedowais312** [comment]
Hi maintainers!  @joe-elliott 
This PR fixes issue #4623 regarding correct parsing of math.MinInt64.  
All local tests pass. Please let me know if any changes are needed!

**Comment by joe-elliott** [comment]
So this is quite a lot of complex code for such an intense edge case. Tests should keep it together, but we still need to weigh the value of the feature against the complexity of the code.

What do you think about an alternative solution like adding constants like `minInt`, `maxInt`? Could we do this in a cleaner way?

**Comment by Syedowais312** [comment]
> So this is quite a lot of complex code for such an intense edge case. Tests should keep it together, but we still need to weigh the value of the feature against the complexity of the code.
> 
> What do you think about an alternative solution like adding constants like `minInt`, `maxInt`? Could we do this in a cleaner way?
Hey @joe-elliott 👋

Thanks for the feedback!

I've updated the PR to use the simpler and cleaner approach you suggested.  
Instead of complex unary parsing logic, I added built-in integer constants:

    minInt → math.MinInt
    maxInt → math.MaxInt

This keeps the lexer minimal, avoids lookahead complexity, and removes the risk of overflow handling code.  
I’ve also updated the tests to use `{ minInt }` and `{ maxInt }`, and all TraceQL tests are passing.

Whenever you get time, please take another look and let me know if anything else needs refinement. Happy to iterate on it!

**Comment by Syedowais312** [comment]
Thanks for the review, @joe-elliott 
implementing `minInt`  /  `maxInt`  directly in the parser makes complete sense. It keeps the lexer simpler and aligns better with how other literals are handled


I'll update the PR shortly with:
-`minInt`  /  `maxInt` evaluate inside the parser
-The corresponding changelog entry
-Updated tests to match the parser-based approach

Thanks again for the guidance - I'll ping you once the new changes are pushed

**Comment by Syedowais312** [comment]
Thanks for the review @joe-elliott!

I’ve updated the PR with the requested changes. The constant handling for `minInt`
and `maxInt` is now implemented cleanly in the parser, directly inside the `integer`
rule as suggested.

All generated parser files were updated using `make gen-traceql-local` and I’ve added
the corresponding tests and CHANGELOG entry.

Local build & full TraceQL test suite also pass successfully.

Please let me know if you’d like any follow-up changes!

File: pkg/traceql/lexer.go [hunk 1]
```diff
@@ -247,6 +253,12 @@ func (l *lexer) Lex(lval *yySymType) int {
 		return tok
 	}
 
+	// builtin integer constants (e.g. minInt, maxInt)
+	if value, ok := builtinIntegerConstants[l.TokenText()]; ok {
```
Comment (joe-elliott) [inline pkg/traceql/lexer.go]: is this code still necessary?

Comment (Syedowais312) [reply pkg/traceql/lexer.go]: Thanks for the follow-up
This block is now no longer needed since `minInt` and `maxInt` are handled directly in the parser via the updated `integer` rule. The lexer no longer needs to check for builtin integer constants

I'll remove this secction and push an update.

Please let me know if you'd like me to include anything else in this cleanup!

**Comment by Syedowais312** [comment]
@joe-elliott 

I’ve moved the minInt/maxInt handling into the parser and removed the redundant lexer logic. The branch is now rebased and ready for another look.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,8 +1,5 @@
 ## main / unreleased
 
-* [CHANGE] Add stricter validation to user-configurable overrides API, such as refusing duplicate labels [#6008](https://github.com/grafana/tempo/pull/6008) (@carles-grafana)
```
Comment (joe-elliott) [inline CHANGELOG.md:3]: deleted changelog entries?

**Comment by Syedowais312** [comment]
Thanks for pointing this out 
I've reset `CHANGELOG.md`  to upstream, restored the removed entries, and added a single changing entry for this TraceQL change under `main / unreleased` .

Please let me know if the wording looks good.

**Comment by joe-elliott** [comment]
Seeing some CI issues

- Don't worry about the linter. I think it's a bug or perhaps a miss configuration on our part. Those are invalid errors.
- The unit test failures need to be corrected. You changed the error return string which is breaking some tests. I prefer your error return so let's keep it and just update the tests to match the new strings.
- The vendor check is due to unexpected differences in the compiled .y file. This might be due to a version difference between what's on your machine and what the CIs using. If you are unable to correct I should be able to rebuild the file and push a commit. `make gen-traceql` should pull a docker container and use the same version as CI.

**Comment by Syedowais312** [comment]
Thanks for the detailed review — this is very helpful.

Acknowledged on all points 👍

• I’ll revert the error return strings back to their original values and update the implementation so existing unit tests continue to pass without modifying expected messages.

• I’ll investigate the vendor diff. Most likely this is coming from a generator/version mismatch on my local setup. I’ll try regenerating via `make gen-traceql` using the same tooling version as CI.  
If I still can’t get a clean diff, I’ll leave the generated file unchanged and only include the minimal necessary changes.

• Understood on the linter errors — I’ll ignore those for now as suggested.

I’ll push an updated commit shortly addressing the test failures and vendor differences.  
Please let me know if there’s anything else you’d like me to be careful about here.

Thanks again for the guidance!

**Comment by joe-elliott** [comment]
> • I’ll revert the error return strings back to their original values and update the implementation so existing unit tests continue to pass without modifying expected messages.

No, please keep your error strings. Update the tests to match.

**Comment by Syedowais312** [comment]
@joe-elliott 
Thanks for the clarification 👍  
I’ve kept the new error strings, updated the affected tests, and regenerated the TraceQL files using the CI image.  
Fixed remaining frontend unit tests to align with updated TraceQL parse error messages.
All relevant unit tests now pass locally.

**Comment by Syedowais312** [comment]
Thanks for the review and guidance! learned alot through parser and test expectation. Looking forward to contribute more
